Fixes #29974 - allow subnet helpers

lzap opened the pull request:
SSIA

lzap committed:
Fixes #29974 - allow subnet helpers

ekohl commented on app/models/smart_proxy.rb:
```diff
@@ -172,6 +172,6 @@ def get_features
 
   class Jail < ::Safemode::Jail
-    allow :id, :name
+    allow :id, :name, :hostname, :hosts_count
```
Can it be a problem that Katello also defines the hostname?
https://github.com/Katello/katello/blob/a9abccc2c10f5f2c1194d564834d131d57b28401/app/models/katello/concerns/smart_proxy_extensions.rb#L336-L338

Either way, after merging this that can be removed.

tbrisker replied:
why do we need the hosts_count?

lzap replied:
I will do a followup PR into Katello.

That `hosts_count` was added by accident, not needed.

lzap replied:
Done: https://github.com/Katello/katello/pull/8748

timogoebel commented on app/models/smart_proxy.rb:
```diff
@@ -172,6 +172,6 @@ def get_features
 
   class Jail < ::Safemode::Jail
-    allow :id, :name
+    allow :id, :name, :hostname, :hosts_count, :setting
```
Does `setting` allow to retrieve any secret settings?

adamruzicka replied:
It allows retrieving settings of a smart proxy feature. I don't think those contain any secrets

ekohl replied:
Should we perhaps expose those explicitly? Features are just strings and exposed on the individual proxies without any authentication so that's IMHO safe. If you're talking about exposed settings from a proxy then those can contain private information and they're only available via an authenticated endpoint (though I'm not aware of any secrets being passed now).

adamruzicka replied:
From my testing machine:
```
SmartProxyFeature.all.each { |f| puts "#{f.feature.name}: #{f.settings}" }
Pulp: {"pulp_url"=>"https://centos7-katello-nightly-stable.example.com/pulp"}
Pulpcore: {"pulp_url"=>"https://centos7-katello-nightly-stable.example.com", "mirror"=>false, "content_app_url"=>"https://centos7-katello-nightly-stable.example.com/pulp/content"}
TFTP: {"tftp_servername"=>nil}
Puppet CA: {"puppet_url"=>"https://centos7-katello-nightly-stable.example.com:8140", "use_provider"=>["puppetca_hostname_whitelisting", "puppetca_http_api"]}
Puppet: {"puppet_url"=>"https://centos7-katello-nightly-stable.example.com:8140", "use_provider"=>["puppet_proxy_puppet_api"]}
Logs: {}
Templates: {}
HTTPBoot: {"http_port"=>8000, "https_port"=>9090}
```

it looks like those don't contain any secrets, but that is no guarrantee that they never will

ekohl replied:
https://github.com/theforeman/smart_proxy_pulp/blob/55ff29f4cfab971f71bc00273c7623ee8dc5af96/lib/smart_proxy_pulp_plugin/pulpcore_plugin.rb#L17-L18 does expose a username/password. It's mostly used in the development workflow if you don't have certificates set up so you won't see it in production, but exposing this is a security risk.

Another thing to consider is that these settings are IMHO internal implementation details. You should always wrap them in high level logic. For example, the puppet server url has a one that also provides compatibility with older releases which didn't expose this:

https://github.com/theforeman/foreman/blob/c3560417748c744d6f64336b1b98102e6a8bb4c1/app/models/concerns/host_common.rb#L95-L100

That particular helper should be exposed rather than raw settings.

:-1: on exposing `setting`.

lzap replied:
Ok how about that, I have implemented whitelisting.

adamruzicka commented on app/models/smart_proxy.rb:
```diff
@@ -100,6 +100,16 @@ def setting(feature, setting)
+    "https_port",
+  ]
+  def safe_setting(feature, setting)
+    raise ::Foreman::Exception.new(N_("Setting is not whitelisted: %s"), ALLOWED_SETTINGS) unless ALLOWED_SETTINGS.include?(setting)
```
This shows what settings are allowed, but not which setting was actually requested
```ruby
::Foreman::Exception.new(N_("Setting is not whitelisted: %s"), ALLOWED_SETTINGS).message
=> "ERF42-4295 [Foreman::Exception]: Setting is not whitelisted: [\"http_port\", \"https_port\"]"
```

lzap replied:
Fixing and adding tests for this.

lzap commented:
Rebased, can I get a quick blessing for this @tbrisker and @ekohl ? This patch is actually needed to have bootdisk working with EFI HTTP BOOT. Added tests so we are safe.

ekohl reviewed (changes requested):
:-1: because I still think explicit wrappers are better. For example, this doesn't look at the feature at all. You will also end up with duplicated logic.

https://github.com/theforeman/foreman/blob/c3560417748c744d6f64336b1b98102e6a8bb4c1/app/models/operatingsystem.rb#L229-L240

That logic block contains a lot of the same things. We've already seen multiple times that users couldn't even manage to debug that error message. If we expose raw settings it'll be even harder. The wrapper can raise a proper error.

lzap commented:
Okay, I've extracted this into two methods. How about now?

lzap commented:
Tests are passing now @ekohl

adamruzicka reviewed (approved):
It exposes all the settings we need in a safe manner without exposing things we don't need. :+1: from me

ekohl reviewed (approved):
Thanks! @tbrisker any last thoughts?

timogoebel commented on app/models/smart_proxy.rb:
```diff
@@ -100,6 +100,14 @@ def setting(feature, setting)
   end
 
+  def httpboot_http_port
+    setting(:HTTPBoot, 'http_port') || raise(::Foreman::Exception.new(N_("HTTP boot requires proxy with httpboot feature and http_port exposed setting")))
```
Do we need to check if the proxy has the appropriate feature?

ekohl replied:
IIRC the data model is that the settings are stored in the Feature model. So having a setting implies the feature. It may be useful to check it separate when you want to raise more explicit exceptions. I can imagine a `FeatureMissingException` or something. However, that's not a requirement for this PR.

lzap replied:
The code actually does this. Method `settings` uses `try`. If you scroll to the right, you will see the error saying either feature is missing or port is not exposed.

adamruzicka commented:
Katello failures seem unrelated, merging

adamruzicka commented:
Thank you @lzap!

lzap commented:
* CP 2.1: https://github.com/theforeman/foreman/pull/7722
